qwerty/ast.hpp

```diff
@@ -2893,6 +2893,13 @@ struct Kernel : ASTNode, HybridObj {
             EvalDimVarExprVisitor evalVisitor(dimvar_values);
             walk(evalVisitor);
         }
+        // Desugar -- want this to be distinct from canonicalization
+        // since we seem to want to type-check after canonicalizing.
+        // This limits which transformations can avoid typechecking
+        {
+            DesugarVisitor desugarVisitor;
+            walk(desugarVisitor);
+        }
```

Comment:
I like your idea of having a separate visitor for this.

Running the visitor here wouldn't cause any functional problems. However, this code runs for both `@classical` and `@qpu` kernels, and in my opinion, we should have different desugar visitors for `@qpu` versus `@classical`.

Do you see `runExtraTypeChecking()` below? I would create something like that called `runDesugaring()` and call that here instead. Then `QpuKernel`  can define `runDesugaring()` to run your desugar visitor. Like this:

https://github.com/gt-tinker/qwerty/blob/c5f66a0c4fb5a17d27d71920f5f6a7d1baa84969/qwerty/ast.hpp#L3116-L3119

See what I'm saying? (And do you agree?)